### PR TITLE
rpc: Update jsonrpsee v0.15.1

### PR DESCRIPTION
**Pull request by lexnv**
This PR updates the `jsonrpsee` crate from v0.14.0 to v0.15.1.

### Jsonrpsee Changes
For full list of changes check: [jsonrpsee/releases](https://github.com/paritytech/jsonrpsee/releases)).

- `Middleware` trait has been split into `HttpMiddleware` and `WsMiddleware` to better capture their differences.
- Return error from subscription callbacks for better ergonomics and removed the `PendingSink` concept (you can still manually accept or reject a sink, but otherwise it'll be accepted automatically if you send a message down it).

// cc @jsdw @niklasad1 

polkadot companion: https://github.com/paritytech/polkadot/pull/5846
comulus companion: https://github.com/paritytech/cumulus/pull/1487

### CODE REVIEW

File: client/beefy/rpc/src/lib.rs [hunk 1]
```diff
@@ -197,9 +199,9 @@ mod tests {
 		let (rpc, _) = setup_io_handler();
 		let request = r#"{"jsonrpc":"2.0","method":"beefy_getFinalizedHead","params":[],"id":1}"#;
 		let expected_response = r#"{"jsonrpc":"2.0","error":{"code":1,"message":"BEEFY RPC endpoint not ready"},"id":1}"#.to_string();
-		let (result, _) = rpc.raw_json_request(&request).await.unwrap();
+		let (response, _) = rpc.raw_json_request(&request).await.unwrap();
 
-		assert_eq!(expected_response, result,);
+		assert_eq!(expected_response, response.result,);
```
Comment (niklasad1) [inline client/beefy/rpc/src/lib.rs:204]:
```suggestion
		assert_eq!(expected_response, response.result);
```

File: client/beefy/rpc/src/lib.rs [hunk 2]
```diff
@@ -135,19 +135,21 @@ impl<Block> BeefyApiServer<notification::EncodedSignedCommitment, Block::Hash> f
 where
 	Block: BlockT,
 {
-	fn subscribe_justifications(&self, pending: PendingSubscription) {
+	fn subscribe_justifications(
+		&self,
+		mut sink: SubscriptionSink,
+	) -> Result<(), SubscriptionEmptyError> {
```
Comment (niklasad1) [inline client/beefy/rpc/src/lib.rs:141]: can't we use `JsonRpseeError` instead?

File: client/rpc/src/author/mod.rs [hunk 3]
```diff
@@ -199,19 +204,16 @@ where
...
 					return
 				},
 			};
 
-			let mut sink = match pending.accept() {
-				Some(sink) => sink,
-				_ => return,
-			};
-
+			// NOTE: the sink is accepted under the hood for ergonomic purposes.
```
Comment (niklasad1) [inline client/rpc/src/author/mod.rs:212]: I think you can stray that note

**Review by niklasad1 — APPROVED** [review]
great LGTM,

as a follow-up we should update the `rpc-api` with the client trait overrides and feature-gate the client stuff such that 
other crates such as subxt can re-use the API definitions from `sc-rpc-api` without having to bring lots needless dependencies such as the `jsonrpsee server stuff`.

**Comment by jsdw** [comment]
> as a follow-up we should update the rpc-api with the client trait overrides and feature-gate the client stuff such that
other crates such as subxt can re-use the API definitions from sc-rpc-api without having to bring lots needless dependencies such as the jsonrpsee server stuff.

Interesting thought; I added an issue in subxt to point at this. Not sure whether we'd want it or not there yet but worth a thought :) https://github.com/paritytech/subxt/issues/605

File: client/rpc/src/state/mod.rs [hunk 4]
```diff
@@ -318,19 +319,26 @@ where
...
+	fn subscribe_storage(
+		&self,
+		mut sink: SubscriptionSink,
+		keys: Option<Vec<StorageKey>>,
+	) -> SubscriptionResult {
 		if keys.is_none() {
 			if let Err(err) = self.deny_unsafe.check_if_safe() {
 				let _ = sink.reject(JsonRpseeError::from(err));
-				return
+				// NOTE: `SubscriptionEmptyError` is introduced for ergonomic purposes.
```
Comment (jsdw) [inline client/rpc/src/state/mod.rs:335]: is this note just pointing out why we return `Result`s now? Probably not needed (we don't bring it up anywhere else either) :)

**Review by jsdw — APPROVED** [review]
LGTM but let's remove the middleware code duplication :)

File: Cargo.lock [hunk 5]
```diff
@@ -1084,68 +1070,65 @@ dependencies = [
...
+checksum = "749d0d6022c9038dccf480bdde2a38d435937335bf2bb0f14e815d94517cdce8"
 dependencies = [
- "cranelift-entity 0.85.0",
+ "cranelift-entity 0.85.3",
 ]
 
 [[package]]
 name = "cranelift-codegen"
-version = "0.76.0"
+version = "0.82.3"
```
Comment (niklasad1) [inline Cargo.lock:1123]: this is unrelated to jsonrpsee, can you revert it and take care of it another PR?

**Review by niklasad1 — CHANGES_REQUESTED** [review]
Can you revert the unrelated changes in `Cargo.lock`? 
Lots of unrelated bumps that may cause regressions.

Would be better to take care of that in another PR

**Review by niklasad1 — APPROVED** [review]
:+1:

**Comment by lexnv** [comment]
bot merge

**Pull request merged**